aws/service.ts

```diff
@@ -247,74 +247,109 @@ function computeImage(
         if (!repository) {
             throw new RunError("Expected a container repository for build image");
         }
 
-        let imageDigest: pulumi.Output<string>;
-        // See if we've already built this.
-        if (imageName && buildImageCache.has(imageName)) {
-            // We got a cache hit, simply reuse the existing digest.
-            // Safe to ! the result since we checked buildImageCache.has above.
-            imageDigest = buildImageCache.get(imageName)!;
-            imageDigest.apply(d =>
-                pulumi.log.debug(`    already built: ${imageName} (${d})`, repository));
-        } else {
-            // If we haven't, build and push the local build context to the ECR repository, wait for
-            // that to complete, then return the image name pointing to the ECT repository along
-            // with an environment variable for the image digest to ensure the TaskDefinition get's
-            // replaced IFF the built image changes.
-            const {repositoryUrl, registryId} = repository!;
-            imageDigest = docker.buildAndPushImage(imageName, container.build, repositoryUrl, repository, async () => {
-                // Construct Docker registry auth data by getting the short-lived authorizationToken from ECR, and
-                // extracting the username/password pair after base64-decoding the token.
-                //
-                // See: http://docs.aws.amazon.com/cli/latest/reference/ecr/get-authorization-token.html
-                if (!registryId) {
-                    throw new RunError("Expected registry ID to be defined during push");
-                }
-                const credentials = await aws.ecr.getCredentials({ registryId: registryId });
-                const decodedCredentials = Buffer.from(credentials.authorizationToken, "base64").toString();
-                const [username, password] = decodedCredentials.split(":");
-                if (!password || !username) {
-                    throw new RunError("Invalid credentials");
-                }
-                return {
-                    registry: credentials.proxyEndpoint,
-                    username: username,
-                    password: password,
-                };
-            });
-            if (imageName) {
-                buildImageCache.set(imageName, imageDigest);
-            }
-            imageDigest.apply(d =>
-                pulumi.log.debug(`    build complete: ${imageName} (${d})`, repository));
-        }
-
-        preEnv.IMAGE_DIGEST = imageDigest;
-
-        return pulumi.all([repository.repositoryUrl, pulumi.all(preEnv)])
-                     .apply(([url, e]) => ({ image: url, environment: e }));
+        return computeImageFromBuild(preEnv, container.build, imageName, repository);
     }
     else if (container.image) {
-        return pulumi.all(preEnv).apply(e => ({ image: imageName, environment: e }));
```

Comment:
Note: this line looks buggy to me (though i preserved its buggyness in my refactoring).  Specifically, we check container.image, but then do not use it.  Instead we set `image: imageName`, instead of `image: container.image`.  Is that intentional @lukehoban ?